### PR TITLE
IGNITE-22994 Refactor subscription to activation/deactivation events …

### PR DESCRIPTION
**Pull request by alex-plekhanov**
…for internal components

Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/managers/encryption/GridEncryptionManager.java [hunk 1]
```diff
@@ -311,6 +311,8 @@ public GridEncryptionManager(GridKernalContext ctx) {
             }
         });
 
+        ctx.internalSubscriptionProcessor().registerGlobalStateListener(this);
```
Comment (Vladsz83) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/encryption/GridEncryptionManager.java:314]: `GridTaskProcessor` is a `GridComponent`. Its `onKernalStart` is already called by `IgniteKernal#start()`, `for (GridComponent comp : ctx){ comp.onKernalStart(active). }`. Do we need to keep one of them? Do we need `implements IgniteChangeGlobalStateSupport` for `GridTaskProcessor`?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/datastructures/DataStructuresProcessor.java [hunk 2]
```diff
@@ -259,6 +259,7 @@ public DataStructuresProcessor(GridKernalContext ctx) {
     /** {@inheritDoc} */
     @Override public void start() {
         ctx.event().addLocalEventListener(lsnr, EVT_NODE_LEFT, EVT_NODE_FAILED);
+        ctx.internalSubscriptionProcessor().registerGlobalStateListener(this);
```
Comment (Vladsz83) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/datastructures/DataStructuresProcessor.java:262]: Let's use a line break

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metastorage/persistence/DistributedMetaStorageImpl.java [hunk 3]
```diff
@@ -279,6 +279,8 @@ public DistributedMetaStorageImpl(GridKernalContext ctx) {
             DistributedMetaStorageUpdateAckMessage.class,
             this::onAckMessage
         );
+
+        ctx.internalSubscriptionProcessor().registerGlobalStateListener(this);
```
Comment (Vladsz83) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/metastorage/persistence/DistributedMetaStorageImpl.java:283]: Let's add `// No-op.` in `onActivate(GridKernalContext kctx)`

File: modules/core/src/main/java/org/apache/ignite/internal/processors/subscription/GridInternalSubscriptionProcessor.java [hunk 4]
```diff
@@ -55,6 +56,9 @@ public class GridInternalSubscriptionProcessor extends GridProcessorAdapter {
      */
     private List<DistributedConfigurationLifecycleListener> distributedConfigurationListeners = new ArrayList<>();
 
+    /** */
+    private final List<IgniteChangeGlobalStateSupport> globalStateListeners = new ArrayList<>();
```
Comment (Vladsz83) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/subscription/GridInternalSubscriptionProcessor.java:60]: Lets make `distributedConfigurationListeners` also `final`

File: modules/core/src/main/java/org/apache/ignite/internal/processors/task/GridTaskProcessor.java [hunk 5]
```diff
@@ -200,6 +200,8 @@ public GridTaskProcessor(GridKernalContext ctx) {
         ctx.io().addMessageListener(TOPIC_TASK_CANCEL, new TaskCancelMessageListener());
         ctx.io().addMessageListener(TOPIC_TASK, new JobMessageListener(true));
 
+        ctx.internalSubscriptionProcessor().registerGlobalStateListener(this);
```
Comment (Vladsz83) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/task/GridTaskProcessor.java:203]: `GridTaskProcessor` is a `GridComponent`. Its `onKernalStart()` and `onKernalStop()` are called by `IgniteKernal`. We call twice `onKernalStart()`. Looks wierd. Should it be either `GridComponent` or `IgniteChangeGlobalStateSupport` but not both?

File: modules/core/src/main/java/org/apache/ignite/internal/managers/encryption/GridEncryptionManager.java (hunk 1 above)
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/encryption/GridEncryptionManager.java:314]: `GridEncryptionManager` doesn't call `onKernalStart` on activation. As for `GridTaskProcessor`,  investigation is required and this activity is out of scope of this ticket.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/task/GridTaskProcessor.java (hunk 5 above)
Comment (alex-plekhanov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/task/GridTaskProcessor.java:203]: In scope of refactoring we shouldn't change logic. It was called twice before changes, perhaps there is a reason for this.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/subscription/GridInternalSubscriptionProcessor.java (hunk 4 above)
Comment (alex-plekhanov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/subscription/GridInternalSubscriptionProcessor.java:60]: Fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metastorage/persistence/DistributedMetaStorageImpl.java (hunk 3 above)
Comment (alex-plekhanov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/metastorage/persistence/DistributedMetaStorageImpl.java:283]: Fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/datastructures/DataStructuresProcessor.java (hunk 2 above)
Comment (alex-plekhanov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/datastructures/DataStructuresProcessor.java:263]: Fixed

**Review by Vladsz83 — APPROVED** [review]
